internal/clientcache/cmd/daemon/command_wrapper.go

```diff
@@ -64,10 +77,56 @@ func (w *CommandWrapper) startDaemon(ctx context.Context) bool {
 	return err == nil || strings.Contains(stdErr.String(), "already running")
 }
 
-// addPersonaInCache runs AddPersonaCommand
-func (w *CommandWrapper) addPersonaInCache(ctx context.Context) bool {
-	c := AddTokenCommand{Command: base.NewCommand(w.ui)}
-	c.Flags()
-	apiErr, err := c.Add(ctx)
+// addTokenToCache runs AddTokenCommand with the token used in, or retrieved by
+// the wrapped command.
+func (w *CommandWrapper) addTokenToCache(ctx context.Context, token string) bool {
+	com := AddTokenCommand{Command: base.NewCommand(w.BaseCommand().UI)}
+	client, err := w.BaseCommand().Client()
+	if err != nil {
+		return false
+	}
+	keyringType, tokName, err := w.BaseCommand().DiscoverKeyringTokenInfo()
+	if err != nil {
+		return false
+	}
+	if token != "" {
+		client.SetToken(token)
+	}
+
+	// Since the daemon might have just started, we need to wait until it can
+	// respond to our requests
+	waitCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+	if err := waitForDaemon(waitCtx); err != nil {
+		// TODO: Print the result of this out into a log in the dot directory
+		return false
+	}
+
+	apiErr, err := com.Add(ctx, client, keyringType, tokName)
 	return err == nil && apiErr == nil
 }
+
+// waitForDaemon continually looks for the unix socket until it is found or the
+// provided context is done. It returns an error if the unix socket is not found
+// before the context is done.
+func waitForDaemon(ctx context.Context) error {
+	dotPath, err := DefaultDotDirectory(ctx)
+	if err != nil {
+		return err
+	}
+	timer := time.NewTimer(0)
+
+	addr := daemon.SocketAddress(dotPath)
```

Comment:
Just a note that it may be nicer to pass around the address as a `*url.URL` and reference e.g. `Path`, depending on how often something like this has to be done.